### PR TITLE
Nanokvm api driver

### PR DESCRIPTION
**Pull request by mangelajo**
## Summary by CodeRabbit

* **New Features**
  * Added NanoKVM driver enabling video capture (snapshots and streaming), keyboard/mouse control, and device management features (reboot, image mounting, system info).

* **Documentation**
  * Added comprehensive NanoKVM driver documentation with installation instructions, configuration examples, and CLI usage guides.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+                        buffer = BytesIO()
+                        frame.save(buffer, format="JPEG")
+                        data = buffer.getvalue()
+                        await send_stream.send(data)
+                else:
+                    self.logger.error(f"Error streaming video: {e}")
+                    raise
+
+        # Start the video streaming task
+        task = asyncio.create_task(stream_video())
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py]: This uses `asyncio.create_task()` directly, but the project convention is to use `anyio` for async concurrency (used throughout the codebase and listed as a dependency). This should use `anyio.create_task_group()` for portability and consistency with other drivers.

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+
+    def __post_init__(self):
+        if hasattr(super(), "__post_init__"):
+            super().__post_init__()
+
+    @classmethod
+    def client(cls) -> str:
+        return "jumpstarter_driver_nanokvm.client.NanoKVMHIDClient"
+
+    async def _reset_client(self):
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py]: `_reset_client()`, `_get_client()`, and `close()` are duplicated verbatim between `NanoKVMVideo` (lines 65-93) and `NanoKVMHID` (lines 189-217). This is ~30 lines of copy-pasted code that could drift out of sync. Consider extracting to a shared base class.

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+            # would require additional implementation in the nanokvm library
+            self.logger.warning("Serial console access not yet fully implemented")
+
+
+    @classmethod
+    def client(cls) -> str:
+        return "jumpstarter_driver_nanokvm.client.NanoKVMClient"
+
+    @export
+    async def get_info(self):
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:332]: `get_info()` manually reimplements the 401 retry logic instead of using the `with_reauth` decorator. The same fragile pattern (defining a local function, passing `video_driver` as `self`) repeats across `reboot()`, `mount_image()`, `download_image()`, `get_mounted_image()`, `get_cdrom_status()`, `is_image_download_enabled()`, `get_image_download_status()`, and `get_images()`. This is inconsistent and fragile -- it works because `with_reauth` accesses `self._get_client()` and `self._reset_client()`, but a refactor could easily break it silently. If the duplicated client code (comment above) were extracted to a base class, these methods could use `with_reauth` directly.

File: python/packages/jumpstarter-driver-nanokvm/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+authors = [
+    { name = "Miguel Angel Ajo", email = "miguelangel@ajo.es" }
+]
+requires-python = ">=3.11"
+dependencies = [
+    "anyio>=4.10.0",
+    "jumpstarter",
+    "jumpstarter-driver-composite",
+    "jumpstarter-driver-pyserial",
+    "nanokvm @ git+https://github.com/mangelajo/python-nanokvm.git@dev",
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/pyproject.toml]: This dependency points to a mutable git branch (`@dev`). The `dev` branch could change at any time, making builds non-reproducible and potentially breaking them without any change in this repo. Consider pinning to a specific commit hash (e.g., `@abc1234`) or using a published release.

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+            dx: Horizontal scroll amount
+            dy: Vertical scroll amount (positive=up, negative=down)
+        """
+        client = await self._get_client()
+        await client.mouse_scroll(dx, dy)
+        self.logger.debug(f"Mouse scrolled: ({dx}, {dy})")
+
+
+@dataclass(kw_only=True)
+class NanoKVMSerial(PySerial):
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py]: `NanoKVMSerial` is a placeholder -- it hardcodes `url = "rfc2217://localhost:2217"` and the `NanoKVM` composite just logs a warning when `enable_serial=True`. This is effectively dead code. Consider removing it from this PR and adding it when the implementation is ready, to avoid users trying to enable a non-functional feature.

File: python/packages/jumpstarter-driver-nanokvm/README.md [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+| ssh_password   | SSH password for serial console            | str   | no       | "root"  |
+| ssh_port       | SSH port for serial console                | int   | no       | 22      |
+
+## Architecture
+
+The NanoKVM driver is a composite driver that provides three main interfaces:
+
+1. **video**: Video streaming and snapshot capture
+2. **hid**: Keyboard and mouse HID control
+3. **serial**: Serial console access (optional, future feature)
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/README.md]: The docs list serial console access as one of the three main interfaces and document config parameters for it (`enable_serial`, `ssh_username`, `ssh_password`, `ssh_port`), but the implementation is a non-functional placeholder. This will mislead users into thinking serial works. Either remove serial references entirely or add a prominent "not yet implemented" disclaimer.

File: python/packages/jumpstarter-driver-nanokvm/README.md [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+
+## API Reference
+
+### NanoKVMClient
+
+```{eval-rst}
+.. autoclass:: jumpstarter_driver_nanokvm.client.NanoKVMClient()
+    :members: get_info, reboot, mount_image, download_image, get_mounted_image, get_cdrom_status, is_image_download_enabled, get_image_download_status
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/README.md]: `get_images` is missing from the `:members:` list. It is exported, tested, and functional but will not appear in the generated API docs.

Suggested fix:
```
:members: get_info, reboot, mount_image, download_image, get_mounted_image, get_cdrom_status, is_image_download_enabled, get_image_download_status, get_images
```

**Review by raballew — COMMENTED** [review]
# Missing test coverage in `driver_test.py`

The following exported methods have no tests at all:
- `mouse_move_rel`
- `mouse_scroll`
- `mouse_click` with x/y coordinates
- `mouse_click` with `MouseButton` enum (only string `"left"` is tested)
- `mount_image`, `download_image`, `get_mounted_image`, `get_cdrom_status`, `is_image_download_enabled`, `get_image_download_status` (all exported via `NanoKVM` composite but untested)
- `with_reauth` error recovery path (the decorator is core logic but never exercised in tests -- a test simulating a 401 followed by successful retry would verify this critical path)
- `snapshot` when the stream yields zero frames (should hit `RuntimeError("No frames available")`)
- `press_key` with multi-char input (driver logs a warning for `len(key) > 2` but this is untested)

Also: `test_nanokvm_hid_press_key` uses `assert_called_with` (checks last call) instead of `assert_called_once_with`, so multiple calls would silently pass.

# No `client_test.py` - project convention gap

Other drivers in this project (power, can, flashers, ridesx) have dedicated `client_test.py` files. The NanoKVM `client.py` has non-trivial logic that warrants its own tests:
- `NanoKVMHIDClient.mouse_click()` has branching logic for `x`/`y` being `None` vs provided
- CLI `paste` command uses `.encode().decode("unicode_escape")` which can raise on invalid input
- CLI `mouse click` command maps string button names to `MouseButton` enums

**Review by raballew — CHANGES_REQUESTED** [review]

File: python/packages/jumpstarter-driver-nanokvm/README.md (hunk 7 above)
Comment (mangelajo) [reply python/packages/jumpstarter-driver-nanokvm/README.md]: done.

File: python/packages/jumpstarter-driver-nanokvm/README.md (hunk 6 above)
Comment (mangelajo) [reply python/packages/jumpstarter-driver-nanokvm/README.md]: True , I removed that incomplete part of the implementation.

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py (hunk 2 above)
Comment (mangelajo) [reply python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py]: Done.

File: python/packages/jumpstarter-driver-nanokvm/pyproject.toml (hunk 4 above)
Comment (mangelajo) [reply python/packages/jumpstarter-driver-nanokvm/pyproject.toml]: It's my own fork, but let me check I think the merged my PRs already

**Comment by mangelajo** [comment]
@ambient-code please rebase this

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+            button_map = {
+                "left": MouseButton.LEFT,
+                "right": MouseButton.RIGHT,
+                "middle": MouseButton.MIDDLE,
+            }
+            button = button_map.get(button.lower(), MouseButton.LEFT)
+
+        client = await self._get_client()
+        await client.mouse_click(button, x, y)
+        self.logger.info(f"Mouse {button.name} clicked")
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:271]: [HIGH] `mouse_click` will crash with `AttributeError` when the `button` argument arrives as a float. The gRPC framework serializes `IntEnum` values as protobuf `number_value`, which deserializes as Python `float` (e.g., `MouseButton.LEFT` becomes `1.0`). The current code only handles `str` conversion, so `(1.0).name` blows up.

Suggested fix: add numeric handling before the string check:
```python
if isinstance(button, (int, float)):
    button = MouseButton(int(button))
```

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+
+    async def _reset_client(self):
+        """Reset the client, forcing re-authentication"""
+        if self._client is not None:
+            try:
+                await self._client.close()
+            except Exception as e:
+                self.logger.debug(f"Error closing client during reset: {e}")
+        self._client = None
+        self._client_ctx = None
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:81]: [HIGH] `_reset_client` calls `await self._client.close()`, but `NanoKVMAPIClient` doesn't implement `close()`. The `AttributeError` gets silently swallowed, and the aiohttp session leaks on every re-authentication.

Suggested fix: replace with `await self._client_ctx.__aexit__(None, None, None)` to properly clean up the context manager.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    "nanokvm>=1.0.0",
+    "aiohttp",
+    "pillow",
+    "pydantic",
+    "yarl",
+    "click",
+]
+
+[project.entry-points."jumpstarter.drivers"]
+NanoKVM = "jumpstarter_driver_nanokvm.driver:NanoKVM"
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/pyproject.toml:24]: [HIGH] Only `NanoKVM` is registered as an entry point. The child drivers `NanoKVMVideo` and `NanoKVMHID` are missing entry points, which means they can't be discovered independently. Other composite drivers in this codebase (dutlink, qemu) register all their children.

Suggested fix: add entry points for `NanoKVMVideo` and `NanoKVMHID`.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+            This uses paste_text under the hood, so it supports the same character set.
+            For special keys like Enter, use '\\n'. For Tab, use '\\t'.
+        """
+        if len(key) > 2:  # Allow for escape sequences like \n, \t
+            self.logger.warning(f"press_key should be used with single characters, got: {key}")
+
+        client = await self._get_client()
+        await client.paste_text(key)
+        self.logger.debug(f"Pressed key: {repr(key)}")
+
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:212]: [HIGH] `press_key()` delegates to `paste_text()`, which only supports printable characters. FR-005 requires modifier key combinations (e.g., Ctrl+Alt+Delete), but the upstream library lacks a raw HID API for this.

Suggested fix: either add a method using the NanoKVM HID endpoint directly, or at minimum document this limitation clearly.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/pyproject.toml [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    "anyio>=4.10.0",
+    "jumpstarter",
+    "jumpstarter-driver-composite",
+    "nanokvm>=1.0.0",
+    "aiohttp",
+    "pillow",
+    "pydantic",
+    "yarl",
+    "click",
+]
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/pyproject.toml:21]: [HIGH] `pyproject.toml` declares `nanokvm>=1.0.0` but has no `[tool.uv.sources]` section to specify the git repository. The lock file resolves version `0.0.1` from git, creating a version/source mismatch. Running `uv lock` fresh will fail.

Suggested fix: add a `[tool.uv.sources]` section with the git source for nanokvm and workspace sources for jumpstarter dependencies.

_AI-generated, human reviewed_

File: python/uv.lock [hunk 13]
```diff
@@ -2710,6 +2711,48 @@ dev = [
...
+    { name = "nanokvm", git = "https://github.com/mangelajo/python-nanokvm.git?rev=dev" },
+    { name = "pillow" },
+    { name = "pydantic" },
+    { name = "yarl" },
+]
+
+[package.metadata.requires-dev]
+dev = [
+    { name = "pytest", specifier = ">=8.3.3" },
+    { name = "pytest-cov", specifier = ">=6.0.0" },
```
Comment (raballew) [inline python/uv.lock:2753]: [HIGH] The lock file is stale and out of sync with pyproject.toml. Issues include: a phantom `jumpstarter-driver-pyserial` entry (not in pyproject.toml), version mismatch for nanokvm (>=1.0.0 vs resolved 0.0.1), and `pytest-asyncio` declared as dev-dependency but absent from the lock file.

Suggested fix: fix the `[tool.uv.sources]` in pyproject.toml first, then regenerate with `uv lock`.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
+"""Tests for NanoKVM driver"""
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py:1]: [HIGH] The `with_reauth` decorator wraps all 14 exported methods and `_is_unauthorized_error` has branching logic, but there are no tests simulating a 401 response or verifying retry behavior.

Suggested fix: add tests that simulate a 401 response followed by successful retry, and unit tests for `_is_unauthorized_error` edge cases.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+__all__ = ["NanoKVMVideo", "NanoKVMHID", "NanoKVM", "MouseButton"]
+
+
+def _is_unauthorized_error(error: Exception) -> bool:
+    """Check if an error is a 401 Unauthorized error"""
+    if isinstance(error, ClientResponseError):
+        return error.status == 401
+    # Also check for string representation in case error is wrapped
+    error_str = str(error)
+    return "401" in error_str and ("Unauthorized" in error_str or "unauthorized" in error_str.lower())
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:25]: [MEDIUM] `_is_unauthorized_error` uses `"401" in error_str` which is fragile substring matching. This can false-positive on unrelated strings containing "401" (e.g., port 4012), triggering unnecessary re-authentication.

Suggested fix: walk the exception chain and check for `ClientResponseError.status == 401`, or use a word boundary regex.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 16]
```diff
@@ -0,0 +1,452 @@
...
+    async def paste_text(self, text: str):
+        """
+        Paste text via keyboard HID simulation
+
+        Args:
+            text: Text to paste (limited character set supported)
+        """
+        client = await self._get_client()
+        await client.paste_text(text)
+        self.logger.info(f"Pasted text: {text}")
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:191]: [MEDIUM] `self.logger.info(f"Pasted text: {text}")` logs the full pasted text at INFO level. HID paste is commonly used for passwords, so this could leak secrets.

Suggested fix: downgrade to `self.logger.debug(f"Pasted text ({len(text)} chars)")`.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/client.py [hunk 17]
```diff
@@ -0,0 +1,536 @@
...
+        @base.command()
+        @click.argument("output", type=click.Path(), default="snapshot.jpg")
+        def snapshot(output):
+            """Take a snapshot and save to file"""
+            image = self.snapshot()
+            image.save(output)
+            click.echo(f"Snapshot saved to {output}")
+            click.echo(f"Image size: {image.size[0]}x{image.size[1]}")
+
+        return base
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/client.py:51]: [MEDIUM] `NanoKVMVideo.stream()` is decorated with `@exportstream` but `NanoKVMVideoClient` has no corresponding `stream()` method. This means ~38 lines of server-side stream code (with duplicated re-auth logic) are unreachable and untested.

Suggested fix: either add a client-side `stream()` method with tests, or remove the server-side stream code until it's needed.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py:2]: [MEDIUM] Eight exported methods have zero test coverage: `mount_image`, `download_image`, `get_mounted_image`, `get_cdrom_status`, `is_image_download_enabled`, `get_image_download_status`, `mouse_move_rel`, and `mouse_scroll`.

Suggested fix: add mock-based tests. Prioritize `mount_image` (has branching logic) and `mouse_move_rel`/`mouse_scroll`.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 19]
```diff
@@ -0,0 +1,452 @@
...
+                host=self.host,
+                username=self.username,
+                password=self.password,
+            ),
+            "hid": NanoKVMHID(
+                host=self.host,
+                username=self.username,
+                password=self.password,
+            ),
+        }
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:314]: [MEDIUM] `NanoKVMVideo` and `NanoKVMHID` each create their own `NanoKVMAPIClient`, doubling resource usage and authentication calls. The plan mentions sharing the HTTP session via the parent, but this isn't implemented.

Suggested fix: share a single `NanoKVMAPIClient` instance across child drivers.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 20]
```diff
@@ -0,0 +1,452 @@
...
+
+        Args:
+            url: URL of the image to download
+
+        Returns:
+            Dictionary with download status information
+        """
+        client = await self._get_client()
+        status = await client.download_image(url=url)
+        self.logger.info(f"Started download from {url}")
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:378]: [MEDIUM] The `url` parameter in `download_image` is passed directly to the NanoKVM download endpoint without scheme validation, which could enable SSRF.

Suggested fix: validate the URL scheme (allow only http/https) and document the SSRF risk.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 21]
```diff
@@ -0,0 +1,452 @@
...
+    }
+
+
+@dataclass(kw_only=True)
+class NanoKVMDriverBase(Driver):
+    """Base class for NanoKVM drivers with shared client management"""
+
+    host: str
+    username: str = "admin"
+    password: str = "admin"
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:64]: [MEDIUM] The `password` field in `NanoKVMDriverBase` doesn't use `repr=False`, so it will appear in default `repr()` output and may leak in logs or tracebacks.

Suggested fix: use `field(repr=False)` for the password field.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+from unittest.mock import AsyncMock, MagicMock, patch
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py:3]: [MEDIUM] Around 20 narration comments in the test file, plus redundant comments in driver.py and client.py. Per project conventions, comments are only acceptable as a last resort when the code itself cannot be made self-explanatory.

Suggested fix: remove the narration comments and let the code speak for itself.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py [hunk 23]
```diff
@@ -0,0 +1,452 @@
...
+        Press a key by pasting a single character
+
+        Args:
+            key: Single character or escape sequence to press (e.g., 'a', 'A', '\\n', '\\t')
+
+        Note:
+            This uses paste_text under the hood, so it supports the same character set.
+            For special keys like Enter, use '\\n'. For Tab, use '\\t'.
+        """
+        if len(key) > 2:  # Allow for escape sequences like \n, \t
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver.py:206]: [LOW] `press_key` length validation uses `if len(key) > 2`, which allows two-character strings without warning. Since decoded escape sequences are single characters, the threshold should be `> 1`.

Suggested fix: change to `if len(key) > 1:`.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+    from nanokvm.models import MouseButton
+
+    hid = NanoKVMHID(host="test.local", username="admin", password="admin")
+
+    with serve(hid) as client:
+        # Click left button
+        client.mouse_click("left")
+
+        # Verify the mock was called
+        mock_nanokvm_client.mouse_click.assert_called_once_with(MouseButton.LEFT, None, None)
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py:205]: [LOW] The `mouse_click` test only covers the no-coordinates branch with a left button. The coordinates branch and non-left button conversions are untested.

Suggested fix: add test cases that pass coordinates and use a non-left button value.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py [hunk 25]
```diff
@@ -0,0 +1,227 @@
...
+
+
+def test_nanokvm_hid_client_creation():
+    """Test that NanoKVMHID returns correct client class"""
+    assert NanoKVMHID.client() == "jumpstarter_driver_nanokvm.client.NanoKVMHIDClient"
+
+
+def test_nanokvm_client_creation():
+    """Test that NanoKVM returns correct client class"""
+    assert NanoKVM.client() == "jumpstarter_driver_nanokvm.client.NanoKVMClient"
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py:179]: [LOW] The client class string tests are near-tautological, just verifying a constant equals itself. They inflate coverage without testing any real behavior.

Suggested fix: replace with import verification tests or tests that exercise actual functionality.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py [hunk 26]
```diff
@@ -0,0 +1,227 @@
...
+    video = NanoKVMVideo(host="test.local", username="admin", password="admin")
+
+    with serve(video) as client:
+        # Get a snapshot
+        image = client.snapshot()
+
+        # Verify it's a PIL Image
+        assert isinstance(image, Image.Image)
+        assert image.size == (640, 480)
+
```
Comment (raballew) [inline python/packages/jumpstarter-driver-nanokvm/jumpstarter_driver_nanokvm/driver_test.py:88]: [LOW] The snapshot `skip_frames` test uses identical mock frames, so the off-by-one bug in skip logic (F012) can't be detected.

Suggested fix: use distinguishable frames in the mock and verify the correct frame is returned.

_AI-generated, human reviewed_

**Review by raballew — CHANGES_REQUESTED** [review]
Also take code rabbit into consideration. Lots of the findings seems correct.

**Comment by kirkbrauer** [comment]
@mangelajo ditto on this PR as well, we should merge or close it.

**Comment by mangelajo** [comment]
Close for now, but will continue at some point..
